src/redis-loader.ts

```diff
@@ -0,0 +1,437 @@
+import { EventEmitter } from 'events'
+import * as DataLoader from 'dataloader'
+import * as invariant from 'invariant'
+import { list as redisCommandList } from 'redis-commands'
+import { ScanStream, ScanStreamOptions } from "./scan-stream"
+import { Redis } from 'ioredis'
+import { redisMethods } from './interfaces'
+import { BatchStats, RedisStats } from './stats';
+
+export type statsLogger = (stats: RedisStats) => void
+
+export interface IRedisLoaderOptions {
+  redis: Redis
+  logger?: statsLogger
+}
+
+export class RedisLoader implements redisMethods, EventEmitter {
+  public stats: RedisStats
+  public redis: Redis
+  private logger: statsLogger
+  private dataLoader: DataLoader<{}, {}>
+
+  constructor({ redis, logger }: IRedisLoaderOptions) {
+    invariant(redis, '"redis" is required')
+    this.logger = logger
+    this.stats = new RedisStats()
+    this.redis = redis
+    this.dataLoader = new DataLoader(commands => this.batchFunction(commands), { cache: false })
+  }
+
+  resetStats() {
+    this.stats = new RedisStats()
+  }
+
+  get _redis() {
+    return this.redis
+  }
+
+  startBatch(commands) {
+    return this.stats.startBatch(commands)
+  }
+
+  endBatch(batchStats, error, response) {
+    this.stats.endBatch(batchStats, error, response)
+    this.logStats()
+  }
+
+  logStats() {
+    if (this.logger) {
+      this.logger(this.stats)
+    }
+  }
+
+  async batchFunction(commands) : Promise<any> {
+    invariant(commands, "no commands to run")
+    const batchStats = this.startBatch(commands)
+    return this.redis.multi(commands).exec().then(
+      response => {
+        this.endBatch(batchStats, null, response)
+        return response.map(([error, data]) => error || data)
+      },
+      error => {
+        if (Array.isArray(error.previousErrors)) {
+          error.message = `${error.message}: ${error.previousErrors.map(e => e && e.message)}`
+        }
+        this.endBatch(batchStats, error, null)
+        throw error
+      }
+    )
+  }
+
+  // Dynamically generated methods
```

Comment:
is there no way to just have this acquired through inheritance or something?

Reply:
But if it's also defined in a separate file why not just import that type and use it?